### ISSUE TITLE
Having trouble with SP.List.getItems(query)

### ISSUE THREAD

**Issue by dylancristy** [bravocg/core#2]
Hi, met you guys at SharePoint Fest DC, and looking forward to getting started using this library.

However, I'm having a problem using the `SP.List.getItems(query)` function.  I'm doing a pretty simple thing and trying to get all the list items:

```
var query = SP.CamlQuery.createAllItemsQuery();
var web = new BRAVO.Core.Web(_spPageContextInfo.webServerRelativeUrl);
var list = web.getListByTitle('My List Name');
var itemColl = list.getItems(query);
```

I get the web and the list just fine, but when I try to get the items I'm getting the following error:

> POST http://mysharepointserver/sites/somesite/_api/Web/Lists(guid'aac92c09-94b5-44e6-b8ab-5176377d6f8f')/getItems 400 (Bad Request)       bravo.min.js:2

From the documentation it seems like `getItems` is supposed to take an `SP.CamlQuery` object, correct?

**Comment by gunjandatta** [comment]
This is from the documentation:
getItems
This method returns a SP.ListItemCollection object, based on the specified query.
Input Parameters
• data – A data object, representing the SP.CamlQuery.
Data Properties
Refer to the SP.CamlQuery REST documentation.

getItemsByFilter
This method returns a SP.ListItemCollection object, based on the specified filter.
Input Parameters
• filter – A string value, representing the OData query operations.
o   Note - This filter prepends “$filter=” to the input parameter

Example:
var web = new BRAVO.Core.Web();       // Get's the current web
var list = web.getListByTitle("My List Name");     // Get the list

// Get all items
var items = list.get_Items();

// Get using the "getItems" method
var items = list.getItems( { query: { __metadata: { type: "SP.CamlQuery" }, ViewXml: "&lt;View&gt;&lt;Query&gt;&lt;Where&gt;&lt;Eq&gt;&lt;FieldRef Name='Title' /&gt;&lt;Value Type='Text'&gt;Test 1&lt;/Value&gt;&lt;/Where&gt;&lt;/Eq&gt;&lt;/Query&gt;&lt;/View&gt;" } });

// Get using the REST OData query operations
var items = list.getItemsByFilter("Title eq 'Test'&top=1");

I tested all of the above examples to ensure they are working w/ the various updates I've been making, and they were all successful.

**Comment by gunjandatta** [comment]
I'll start updating the documentation to include examples.

**Comment by dylancristy** [comment]
I guess I am still confused as to the correct usage.

I tried

```
var query = { query: { __metadata: { type: "SP.CamlQuery" }, ViewXml: "<View><Query></Query></View>" } };
var itemColl = list.getItems(query);
```

and it just hangs my browser.  Is that not correct?

Ultimately I am trying to use a CAML query because I want to get all items, but only a subset of fields, so I figured I'd use the `<ViewFields>` part of the query to get that.

I don't think I can do that using the oData filter method.

**Comment by gunjandatta** [comment]
Can you run this script in your console window:

---

// Get the current web
var web = new BRAVO.Core.Web();

// Get the list
var list = web.getListByTitle("Test");

// Create if it doesn't exist
if(!list.exists) {
  // Create the list
  list = web.addList({ BaseTemplate: 100, Title: "Test" });

  // Add a field
  list.addFieldAsXml('&lt;Field ID="{AA3AF8EA-2D8D-4345-8BD9-6017205F6212}" Name="TestValue" StaticName="TestValue" DisplayName="Value" Type="Text" /&gt;');

  // Create items
  for(var i=0; i<10; i++) { 
    var item = list.addItem({ Title: "Test Item " + i, TestValue: i + "" });
  }

}

// Query the items
var items = list.get_Items();

var itemsByCAML = list.getItems( { query: { __metadata: { type: "SP.CamlQuery" }, ViewXml: "&lt;View&gt;&lt;Query&gt;&lt;/Query&gt;&lt;ViewFields&gt;&lt;FieldRef Name='Title' /&gt;&lt;FieldRef Name='TestField' /&gt;&lt;/ViewFields&gt;&lt;/View&gt;" } });

// Log to the console
console.log("Done");

---

Notice that both objects contain the "TestValue" field, even though the default view does not contain it. You can also use the odata filter:

getItemsByFilter("ID gt 0&$select='[Fields you want]'");

Let me know if you this clears up the issue and I'll close this out.

**Comment by dylancristy** [comment]
Haven't run that test script yet, but I was playing around with some other things.  There is definitely something funky going on with `getItems(query)`.

For comparison, I tried using the BRAVO core model, I tried using the regular JSOM model, and I tried using REST but coding up the jQuery Ajax request by hand.  Keep in mind I'm querying a list with over 4K items.

If I set a `RowLimit` of 100 on the query, regular JSOM and regular REST return almost instantly, and if I am using `ViewFields` I get back the expected fields.  Using BRAVO core, it takes several (7-10) seconds to get just 100 items, and it does not respect the `ViewFields` (it returns all fields).

I also tried removing the `RowLimit`, and even retrieving the full 4K items only took about 3 seconds using regular REST (and still respected the `ViewFields`).  Using BRAVO core with no limit basically takes so long that Chrome decides the window is unresponsive and tries to kill it.

So... those are my findings.

Thanks for the tip on how to set up the odata filter string, I will try that out tomorrow!

**Comment by gunjandatta** [comment]
That was going to be my next question. How large of a list are we dealing with. If you are going to execute a request like this, I'd recommend setting the asyncFl to true, and waiting for it to execute so the browser doesn't lock up.

The library is just a helper class to execute REST calls, so it theoretically shouldn't be different than the regular REST execution. I'll take a look at the code surrounding the "results". I know I was doing something w/ the results, so I bet it's getting stuck there.

Thanks for the help on this.

**Comment by gunjandatta** [comment]
I commented out the piece of code that was parsing the results, and will update the latest code now. Please let me know if the getItems method is not running as fast as the direct REST api call. There should be no difference now.

Make sure you clear you cache to ensure you are running the latest.

**Comment by chapstickbomber** [comment]
I noticed the delay on 100 items there as well. What do we lose from
dropping the parsing? Is there a lightweight way to get the same effect?

**Comment by gunjandatta** [comment]
I thought about this last night, and will create an "ensureAllProperties" method which will take the uri of the item, query it and ensure all object properties are there. I'll provide some examples of the properties it was adding by default, but honestly it should be a method and not doing this automatically for collections.

**Comment by gunjandatta** [comment]
For list items using the getItems or getItemsByFilter method, these properties were missing:
AttachmentFiles
ContentType
FieldValuesAsHtml
FieldValuesAsText
FieldValuesForEdit
File
FirstUniqueAncestorSecurableObject
Folder
GetDlpPolicyTip
ParentList
RoleAssignments

I'll add the "ensureAllProperties" method to still allow the developer to get all properties, but it shouldn't be in there by default.

**Comment by gunjandatta** [comment]
Actually, further research into this shows that these "properties' are the "collections" shown at the top of the REST xml. I've updated the code to ensure the result properties are updated and look like the root object.

I've tested this against a list w/ 4100 items in it. A call to query the top 100 items returned w/in a second.

Thank you for testing this out and finding this bug. I've also added the ability to get a "List" object directly from the core class:

var list = new BRAVO.Core.List("Test List");

It follows the same input as the others (hostUrl and asyncFl are optional).
